### PR TITLE
feat: make project optional

### PR DESCRIPTION
**Pull request by cristiand391**
**DEPENDS ON https://github.com/forcedotcom/packaging/pull/638/files**

### What does this PR do?
https://github.com/forcedotcom/cli/discussions/2932

Allow the following commands to be run outside an sfdx project:
* `package delete`
* `package update`
* `package version delete`
* `package version displayancestry`
* `package version list`
* `package version promote`
* `package version report`

### What issues does this PR fix or reference?
@W-16286213@

### CODE REVIEW

File: src/commands/package/version/list.ts [hunk 1]
```diff
@@ -105,7 +106,7 @@ export class PackageVersionListCommand extends SfCommand<PackageVersionListComma
   public async run(): Promise<PackageVersionListCommandResult> {
     const { flags } = await this.parse(PackageVersionListCommand);
     const connection = flags['target-dev-hub'].getConnection(flags['api-version']);
-    const project = SfProject.getInstance();
```
Comment (cristiand391) [inline src/commands/package/version/list.ts:108]: this command didn't have `requiresProject = true` but this line was enforcing it anyway.

File: src/utils/getProject.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * Get the sfdx project from the current dir.
+ * It will return `undefined` if there's no project.
+ * */
+export async function maybeGetProject(): Promise<SfProject | undefined> {
+  try {
+    return await SfProject.resolve();
+  } catch (err) {
+    if (err instanceof Error && err.name === 'InvalidProjectWorkspaceError') {
+      return undefined;
+    }
```
Comment (mshanemc) [inline src/utils/getProject.ts:20]: any swallow all other errors? 
What's the logic of checking the error name then?

**Review by mshanemc — COMMENTED** [review]
why does package:version:update need a Project?  It doesn't **look** like it's reading the project (flags for tag, branch,  etc passed as inputs to the method).

I'm going to ask Vivek about the sfdx-projecet modification commetns.

File: src/utils/getProject.ts (hunk 2 above)
Comment (cristiand391) [reply src/utils/getProject.ts]: > What's the logic of checking the error name then?

none, I think I was just following the error msg from `SfProject.resolve`'s jsdoc. 
Changed to make it swallow all errors that could come from trying to resolve a project.

**Comment by VivekMChawla** [comment]
> why does package:version:update need a Project? It doesn't **look** like it's reading the project (flags for tag, branch, etc passed as inputs to the method).
> 
> I'm going to ask Vivek about the sfdx-projecet modification commetns.

@mshanemc - Perhaps `package version update` wants to update `versionName` or `versionDescription` in the relevant `packageDirectory` object inside `sfdx-project.json`? This would only happen if the `--version-name` or `--version-description` flags were used with the command.

I'm not sure if this command actually DOES work this way or not when inside a project.  I'll test it and report back.

**Review by mshanemc — APPROVED** [review]

**Comment by mshanemc** [comment]
qa notes

outside a project (from this repo)
✅  package version list 
✅ `package version report -p 04tKY000000HDvHYAW`
✅ `package version report -p foo`

File: src/commands/package/version/list.ts [hunk 3]
```diff
@@ -135,7 +136,7 @@ export class PackageVersionListCommand extends SfCommand<PackageVersionListComma
 
       records.forEach((record) => {
         const ids = [record.Id, record.SubscriberPackageVersionId];
-        const aliases = ids.map((id) => project.getAliasesFromPackageId(id)).flat();
+        const aliases = ids.map((id) => (project ? project.getAliasesFromPackageId(id) : id)).flat();
```
Comment (mshanemc) [inline src/commands/package/version/list.ts:139]: where that alias bug is happening.  I think if there is no Project, there should be no aliases

**Comment by mshanemc** [comment]
QA cont

✅ `package version delete` does
✅ `package version promote` tries to (I chose one without code coverage, so it throws, but not because of the project and it IS able to find the package version)

**Pull request merged**